flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/tasks/InputSplitProvider.java

```diff
@@ -35,4 +35,16 @@
 	 *         task shall not consume any further input splits.
 	 */
 	InputSplit getNextInputSplit();
+
+	/**
+	 * Starts the input split provider with a user code class loader.
+	 *
+	 * @param userCodeClassLoader User code class loader to use by the input split provider
+	 */
+	void start(ClassLoader userCodeClassLoader);
```

Comment:
The `start` method's intention is to pass in the `userCodeClassLoader` which is currently created in `Task#run` method. There are two other ways to solve the problem. Either creating the user code class loader outside of `Task` where the `InputSplitProvider` is created or to pass the user code class loader via the `getNextInputSplit` method call to the input split provider.

For the first approach: Creating the user code class loader is a blocking operation, so this would have to executed in a future and upon completion we could create the `Task` instance in the `TaskManager`.


For the second approach: We would have to touch more code but I think everywhere were the `getNextInputSplit` method is called, we have access to the user code class loader.